Adding support for RV32A atomic operations (AMO*, LR, and SC)

blu-peter opened the pull request:
These changes add RV32A atomic operations support.

I'm not sure if this is the best way to manage ISA extensions - whether these should be included within `insn.py`, or whether it would be better to create extension specific instruction files, e.g. `insn_a.py`.

wallento commented:
Thanks a lot, it looks good to me. The actual `execute` implementation is probably rather simple, or are you also thinking towards some kind of parallelism in the model (multi-hart simulation)?

blu-peter commented:
I've added 'execute' implementations for LR, SC, and the AMO* instructions. I've also added a test case for a basic LR/SC sequence, and fixed some issues that I saw with pytest - so both tests pass again (ADDI & LR/SC).

I'd like to add tests for AMO* instructions, but I'm not sure how I would preload state into registers/memory banks. This I think is probably a longer task with some more wide reaching changes - and I'm sure you will have an opinion on how best it would be accomplished.

:)

blu-peter commented:
I should note, as you observed, the implementation only considers a single HART and single stage pipeline. If the model was to be changed to include multiple HARTs, you'd need to have a second layer of lock management on shared memory resources. Also if the model was to use a multi-stage pipeline with interrupt support, the AMO* implementations would need to perform additional checks for reservation integrity.
